packages/paymasters/src/TokenPaymasterProvider.ts

```diff
@@ -0,0 +1,129 @@
+import BN from 'bn.js'
+import { RelayClient, RelayProvider, GSNUnresolvedConstructorInput } from '@opengsn/provider'
+import { PrefixedHexString, toChecksumAddress } from 'ethereumjs-util'
+import { Address, removeHexPrefix, Web3ProviderBaseInterface } from '@opengsn/common/dist'
+import { GSNConfig } from '@opengsn/provider/dist/GSNConfigurator'
+import {
+  PermitERC20UniswapV3PaymasterInstance,
+  PermitInterfaceDAIInstance,
+  PermitInterfaceEIP2612Instance
+} from '../types/truffle-contracts'
+import { RelayRequest } from '@opengsn/common/dist/EIP712/RelayRequest'
+import {
+  PERMIT_SIGHASH_DAI, PERMIT_SIGHASH_EIP2612,
+  PERMIT_SIGNATURE_DAI,
+  PERMIT_SIGNATURE_EIP2612,
+  signAndEncodeDaiPermit,
+  signAndEncodeEIP2612Permit
+} from './PermitPaymasterUtils'
+import { constants } from '@opengsn/common/dist/Constants'
+import { EIP712Domain, EIP712DomainType, EIP712DomainTypeWithoutVersion } from '@opengsn/common/dist/EIP712/TypedRequestData'
+import { TokenPaymasterInteractor } from './TokenPaymasterInteractor'
+
+export interface TokenPaymasterConfig extends GSNConfig {
+  tokenAddress: Address
+  permitAmount?: string | number | BN
+}
```

Comment:
Well, this does not make sense to me. First you add `tokenPaymasterAddress` and `tokenPaymasterDomainSeparators` to the `GSNConfig` and then you inherit from `GSNConfig` and add two more fields. This is two bad solutions used at once. 

I think if we want the `TokenPaymasterProvider` to be a first-class Provider then its configuration is part of the GSNConfig. We may need to find a better arrangement later.

Reply:
Well we have a bit of an issue, since it sits in paymasters package, but the original RelayClient, RelayProvider and GSNConfig sit in other packages, which cannot depend on paymasters. The issue with EIP712Domain is that it's difficult to build it from the token as it may vary between networks, even for the same token. That's why I have it as config param.